server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,392 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Hasura.Eventing.ScheduledTrigger
+  ( processScheduledQueue
+  , runScheduledEventsGenerator
+  ) where
+
+import           Control.Concurrent              (threadDelay)
+import           Control.Exception               (try)
+import           Data.Has
+import           Data.Time.Clock
+import           Hasura.Eventing.HTTP
+import           Hasura.Prelude
+import           Hasura.RQL.DDL.Headers
+import           Hasura.RQL.Types.ScheduledTrigger
+import           Hasura.RQL.Types
+import           Hasura.SQL.DML
+import           Hasura.SQL.Types
+import           System.Cron
+import           Hasura.HTTP
+import           Hasura.Server.Version         (HasVersion)
+
+import qualified Data.Aeson                      as J
+import qualified Data.Aeson.Casing               as J
+import qualified Data.Aeson.TH                   as J
+import qualified Data.TByteString                as TBS
+import qualified Data.Text                       as T
+import qualified Database.PG.Query               as Q
+import qualified Hasura.Logging                  as L
+import qualified Network.HTTP.Client             as HTTP
+import qualified Network.HTTP.Types              as HTTP
+import qualified Text.Builder                    as TB (run)
+import qualified Data.HashMap.Strict             as Map
+
+import           Debug.Trace
+
+invocationVersion :: Version
+invocationVersion = "1"
+
+oneSecond :: Int
+oneSecond = 1000000
+
+oneMinute :: Int
+oneMinute = 60 * oneSecond
+
+oneHour :: Int
+oneHour = 60 * oneMinute
+
+type ScheduledEventPayload = J.Value
+
+scheduledEventsTable :: QualifiedTable
+scheduledEventsTable =
+  QualifiedObject
+    hdbCatalogSchema
+    (TableName $ T.pack "hdb_scheduled_events")
+
+data ScheduledEventSeed
+ = ScheduledEventSeed
+ { sesName          :: !TriggerName
+ , sesScheduledTime :: !UTCTime
+ } deriving (Show, Eq)
+
+-- | ScheduledEvents can be "partial" or "full"
+-- Partial represents the event as present in db
+-- Full represents the partial event combined with schema cache configuration elements
+data SE_P = SE_PARTIAL | SE_FULL
+
+type family Param (p :: k) x
+
+data ScheduledEvent (p :: SE_P)
+  = ScheduledEvent
+  { seId            :: !Text
+  , seName          :: !TriggerName
+  , seScheduledTime :: !UTCTime
+  , seTries         :: !Int
+  , seWebhook       :: !(Param p T.Text)
+  , sePayload       :: !(Param p J.Value)
+  , seRetryConf     :: !(Param p RetryConfST)
```

Comment:
I am okay to do anything. Though, there is also `ScheduledEventSeed`. Because of this I wanted to reduce naming congestion and make it all under one.

Having said that, it would be cool to put `ScheduledEventSeed` in this but I couldn't figure out how to define the `Param` instance for three kinds. Do I need to explicitly write it for each record field type in `ScheduledEvent` ?

Reply:
I think having three separate datatypes (some of which may contain the others) is nicer than getting fancy with type families here (since I’m not sure the type families save us much), but I’m open to being convinced otherwise.